src/tezWrapper.ml

```diff
@@ -0,0 +1,279 @@
+  ledger_issue (st, tez_token_id, addr, tez_to_mutez_nat amnt)
+
+let[@inline] ledger_withdraw_tez_token
+    (st, addr, amnt: fa2_state * Ligo.address * Ligo.tez) : fa2_state =
+  ledger_withdraw (st, kit_token_id, addr, tez_to_mutez_nat amnt)
+
+(*****************************************************************************)
+(**                        {1 FA2 ENTRYPOINTS}                               *)
+(*****************************************************************************)
+
+(* TODO: START COPY-PASTE AND MANUALLY EDIT *)
+let[@inline] tez_wrapper_fa2_get_balance (st, owner, token_id: fa2_state * Ligo.address * fa2_token_id): Ligo.nat =
+  let ledger = st.ledger in
+  let key = (token_id, owner) in
+  let () = if token_id = tez_token_id then () else failwith "FA2_TOKEN_UNDEFINED" in
+  get_fa2_ledger_value ledger key
+
+let[@inline] tez_wrapper_fa2_run_balance_of (st, xs: fa2_state * fa2_balance_of_request list)
+  : fa2_balance_of_response list =
+  List.map
+    (fun (req: fa2_balance_of_request) ->
+       let { owner = owner; token_id = token_id; } : fa2_balance_of_request = req in
+       let blnc = tez_wrapper_fa2_get_balance (st, owner, token_id) in
+       { request=req; balance = blnc; }
+    )
+    xs
+(* TODO: END COPY-PASTE AND MANUALLY EDIT *)
+
+let[@inline] balance_of (state: tez_wrapper_state) (param: fa2_balance_of_param) : LigoOp.operation list * tez_wrapper_state =
+  let _ = ensure_no_tez_given () in
+  let { requests = requests; callback = callback; } = param in
+  let response = tez_wrapper_fa2_run_balance_of (state.fa2_state, requests) in (* TODO: Uses specialized tez_wrapper_fa2_run_balance_of *)
+  let op = LigoOp.Tezos.fa2_balance_of_response_transaction response (Ligo.tez_from_literal "0mutez") callback in
+  ([op], state) (* unchanged state; no attempt to originate vaults either *)
+
+(* TODO: START COPY-PASTE AND MANUALLY EDIT *)
+let[@inline] reverse_op_list (ops: LigoOp.operation list) : LigoOp.operation list =
+  Ligo.List.fold_left
+    (fun ((ops, op) : LigoOp.operation list * LigoOp.operation) -> (op :: ops))
+    ([] : LigoOp.operation list)
+    ops
+
+(* TODO: reverse the output operation list *)
+let[@inline] tez_wrapper_fa2_run_transfer (st, xs: tez_wrapper_state * fa2_transfer list) : tez_wrapper_state * LigoOp.operation list =
+  let state, rev_ops =
+    Ligo.List.fold_left
+      (fun (((st, ops), tx): (tez_wrapper_state * LigoOp.operation list) * fa2_transfer) ->
+         let { from_ = from_; txs = txs; } = tx in
+
+         (* Origination of the from_ vault, if needed *)
+         let op_opt, st, from_vault_address = find_vault_address st from_ in
+         let ops = match op_opt with
+           | None -> ops (* vault is already originated *)
+           | Some origination -> (origination :: ops) in (* originate now *) (* NOTE: reversed! *)
+
+         (* TODO: Do we really need to restrict the to_ and from_ here? *)
```

Comment:
What I worry about with this is that we could end up causing vaults to be created for vaults if someone transfers tez-tokens to a vault address. I'm not sure if this causes vulnerabilities, but it feels a bit odd.

e.g.

> I transfer to address of Vault A -> No record exists in ledger or Vault A -> Vault for Vault A is originated and transfer is executed

At that point the tokens are lost to the vault since we can't issue transfers from Vault A directly, though I guess the end result is just that the tokens are lost (not much different than sending tokens to an unknown address)

Reply:
Ah, I see what you mean. Yes, that could happen but indeed it's not much different than sending tokens to an unknown address. I'm sure that it cannot be used for an attack since we're using bigmaps (the performance of the tez wrapper will not degrade no matter how many vaults someone can create), but I guess it is a little weird. On the other hand, adding the check will raise gas costs, but people will still be able to create vaults for useless accounts if they want so the problem is not entirely gone. :thinking:

Reply:
Hmmm, looking more into this, I think that implementing a check about whether an address belongs to a vault is more complicated than it seems at first glance: the big map is indexed by user address, so figuring out whether an address is a vault address cannot be done unless we keep track of them in a separate map. I added a check for `!Ligo.Tezos.self_address` for now, but we might have to reconvene about the vault check.

Reply:
Ah true, I hadn't thought about that. Given that we're trying to avoid gas costs I agree that it might make more sense to prioritize that over user-friendliness here since allowing users to create extra vault contracts shouldn't affect the functionality of the wrapper.